src/analysis/retail/priest/shadow/modules/tier/Tier31ShadowPriest4Set.tsx

```diff
@@ -0,0 +1,191 @@
+    );
+
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(SPELLS.SHADOW_WORD_PAIN),
+      this.onShadowWordPainCast,
+    );
+
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(TALENTS.SHADOW_CRASH_TALENT),
+      this.onShadowCrashCast,
+    );
+
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.SHADOW_PRIEST_TIER_31_4_SET_BUFF),
+      this.onBuffRemoved,
+    );
+  }
+
+  onBuffApplied(Event: ApplyBuffEvent) {
+    this.timestampOfLast = Event.timestamp;
+    this.stacks = 1;
+    this.recentApplied = true;
+  }
+
+  onBuffStack(Event: ApplyBuffStackEvent) {
+    this.timestampOfLast = Event.timestamp;
+    this.stacks = Event.stack;
+  }
+
+  onBuffRefresh(Event: RefreshBuffEvent) {
+    this.timestampOfLast = Event.timestamp;
+  }
+
+  onBuffRemoved(Event: RemoveBuffEvent) {
+    this.recentStacks = this.stacks;
+    if (MAX_DURATION <= Event.timestamp - this.timestampOfLast) {
+      this.recentStacks = 0;
+    }
+    this.stacks = 0;
+  }
+
+  onShadowCrashCast() {
+    this.castSC = true;
+    //SC hits multiple targets, but we only want to count the stacks used once.
+  }
+
+  onShadowCrash(Event: DamageEvent) {
+    //Since the damage event of the inital hit occurs right after the cast, this will only allow that damage event to be added and increased and not further ticks of the dot.
+    //In addition, the recentStacks is not reset until a new stack is gained. This way, if two shadow Crashes (or shadow word pains) are cast without gaining the buff since the last cast, the damage will not be counted.
+    //Since Shadow crash can damage multiple targets, if a damage event occurs at the same time as another, then it's damage would also be increased, but we do not count the stacks again.
+    if (this.recentSCTimestamp === Event.timestamp || this.recentApplied) {
+      this.recentApplied = false;
+      if (this.castSC) {
+        this.stacksSC += this.recentStacks;
+        this.castSC = false;
+        this.recentSCTimestamp = Event.timestamp;
+      }
+      this.damageSC += calculateEffectiveDamage(Event, this.recentStacks * this.bonusSC);
+    }
+  }
+
+  onShadowWordPainCast() {
+    this.castSWP = true;
+    //Only the initial damage of SWP is increased
+  }
+
+  onShadowWordPain(Event: DamageEvent) {
+    //Since the damage event of the inital hit occurs right after the cast, this will only allow that damage event to be added and increased and not further ticks of the dot.
+    //In addition, the recentStacks is not reset until a new stack is gained. This way, if two shadow word pains are cast (or shadow crashes) without gaining the buff since the last cast, the damage will not be counted.
+    if (this.castSWP && this.recentApplied) {
+      this.recentApplied = false;
+      this.stacksSWP += this.recentStacks;
+      this.damageSWP += calculateEffectiveDamage(Event, this.recentStacks * this.bonusSWP);
+      this.castSWP = false;
+    }
+  }
```

Comment:
It's likely you could detect this instead by looking for SWP damage that doesn't have the `tick = true` flag, this indicating it was the direct initial hit. I can't confirm this unfortunately because all the PTR logs I could find were consuming the proc with Shadow Crash instead of a SWP hardcast.

Reply:
That is a much better way to do it.  
I've been using this log /report/61fWqPBYTDx3M8rX/6-Normal+Gnarlroot+-+Kill+(2:54)/Publik/standard/statistics
It uses SW:P twice to spend stacks, and using tick gave the same results as checking when it was cast.

Reply:
Oh you even referenced the log in the PR lol, I should have used my brain